public/pages/utils/anomalyResultUtils.ts

```diff
@@ -483,3 +523,274 @@ export const parsePureAnomalies = (
   }
   return anomalies;
 };
+
+export type FeatureDataPoint = {
+  isMissing: boolean;
+  plotTime: number;
+  startTime: number;
+  endTime: number;
+};
+
+export const getFeatureDataPoints = (
+  featureData: FeatureAggregationData[],
+  interval: number,
+  dateRange: DateRange | undefined
+): FeatureDataPoint[] => {
+  const featureDataPoints = [] as FeatureDataPoint[];
+  if (!dateRange) {
+    return featureDataPoints;
+  }
+  const existingTimes = isEmpty(featureData)
+    ? []
+    : featureData
+        .map(feature => feature.endTime)
+        .filter(featureTime => featureTime != undefined);
+  for (
+    let currentTime = dateRange.startDate;
+    currentTime <
+    getFloorPlotTime(dateRange.endDate - interval * MIN_IN_MILLI_SECS);
+    currentTime += interval * MIN_IN_MILLI_SECS
+  ) {
+    const isExisting = findTimeExistsInWindow(
+      existingTimes,
+      currentTime,
+      currentTime + interval * MIN_IN_MILLI_SECS
+    );
+    featureDataPoints.push({
+      isMissing: !isExisting,
+      plotTime: currentTime + interval * MIN_IN_MILLI_SECS,
+      startTime: currentTime,
+      endTime: currentTime + interval * MIN_IN_MILLI_SECS,
+    });
+  }
+
+  return featureDataPoints;
+};
+
+const findTimeExistsInWindow = (
+  timestamps: any[],
+  startTime: number,
+  endTime: number
+): boolean => {
+  const result = timestamps.filter(
+    timestamp => timestamp >= startTime && timestamp <= endTime
+  );
+  return !isEmpty(result);
+};
+
+const sampleFeatureMissingDataPoints = (
+  featureMissingDataPoints: FeatureDataPoint[],
+  dateRange?: DateRange
+): FeatureDataPoint[] => {
+  if (!dateRange) {
+    return featureMissingDataPoints;
+  }
+  const sampleTimeWindows = calculateSampleWindowsWithMaxDataPoints(
+    MAX_FEATURE_ANNOTATIONS,
+    dateRange
+  );
+
+  const sampledResults = [] as FeatureDataPoint[];
+  for (const timeWindow of sampleTimeWindows) {
+    const sampledDataPoint = getMiddleDataPoint(
+      getDataPointsInWindow(featureMissingDataPoints, timeWindow)
+    );
+    if (sampledDataPoint) {
+      sampledResults.push({
+        ...sampledDataPoint,
+        startTime: Math.max(timeWindow.startDate, sampledDataPoint.startTime),
+        endTime: Math.max(timeWindow.endDate, sampledDataPoint.endTime),
+      } as FeatureDataPoint);
+    }
+  }
+
+  return sampledResults;
+};
+
+const getMiddleDataPoint = (arr: any[]) => {
+  if (arr && arr.length > 0) {
+    return arr[Math.floor(arr.length / 2)];
+  }
+  return undefined;
+};
+
+const getDataPointsInWindow = (
+  dataPoints: FeatureDataPoint[],
+  timeWindow: DateRange
+) => {
+  return dataPoints.filter(
+    dataPoint =>
+      get(dataPoint, 'plotTime', 0) >= timeWindow.startDate &&
+      get(dataPoint, 'plotTime', 0) < timeWindow.endDate
```

Comment:
is there a reason you filter out by just `<` end date here, and filter out by `<=` end date in other places?

Reply:
that's good catch. using `<=` is a bug which I will fix in next commit.